### PR TITLE
Latent metric anomaly detection

### PR DESCRIPTION
**Pull request by ndeutschmann**
_This is a remake of #24 with signed-off commits_

This was work in most part performed by @fcraighero while at IBM Research Zurich. It implements three algorithms for epistemic uncertainty estimation/anomaly detection based on proxies for data density in the latent space of neural networks.

It provides the following new posthocUQ methods:

- `uq360.algorithms.layer_scoring.knn.KNN`
- `uq360.algorithms.layer_scoring.aklpe.AKLPE`
- `uq360.algorithms.layer_scoring.mahalanobis.Mahalanobis`

The first two methods are based on k-Nearest-Neighbor searches, for which we provide three solutions
- `uq360.utils.transformers.nearest_neighbors.exact.ExactNearestNeighbors` based on `sklearn`
- `uq360.utils.transformers.nearest_neighbors.pynndescent.PyNNDNearestNeighbors` based on `pynndescent`
- `uq360.utils.transformers.nearest_neighbors.faiss.FAISSNearestNeighbors` based on `faiss`

Still TODO: implementing tests (code has been checked of course though). Opening this PR to allow discussions.

_NB: Because I merged the repo IBM/DANTE into UQ360 to preserve history, adding the signoff required some gymnastics.  I'm therefore using a second branch_

### CODE REVIEW

**Comment by ndeutschmann** [comment]
Tests are done with essentially full coverage, I'm therefore officially opening the PR

**Comment by pronics2004** [comment]
Thanks @ndeutschmann! This is looking great!

If it is possible it will be good to add a notebook in the `examples` folder showing the usage for simple cases.

Also, could you edit this file so that the new methods show in the `apidocs` (https://uq360.readthedocs.io/) https://github.com/IBM/UQ360/blob/main/docs/extrinsic.rst

**Comment by ndeutschmann** [comment]
And the docs are now up to date! Feel free to let me know if anything else should be addressed.

**Review by pronics2004 — APPROVED** [review]
LGTM!

**Comment by pronics2004** [comment]
> And the docs are now up to date! Feel free to let me know if anything else should be addressed.

We can merge this PR, if there are no other updates.

FYI: @jirin1a @SoumyaTGhosh

**Comment by ndeutschmann** [comment]
Awesome, thanks! Let me just delete the `.py` copy of the examples notebook and we're ready to go!

**Comment by ndeutschmann** [comment]
Ready to go!
